Scripts and tools: Bump flake8 to 3.7.8

Empact opened the pull request:
This is a second go at #15221, fixing new lints in:
W504 line break after binary operator
W605 invalid escape sequence
F841 local variable 'e' is assigned to but never used

This time around:
* One commit per rule, for easier review
* I went with the PEP-8 style of breaking before binary operators
* I looked into the raw regex newline issue, and found that raw strings with newlines embedded do work appropriately. E.g. run `re.match(r" \n ", " \n ")` to check this for yourself. `re.MULTILINE` exists to modify `^` and `$` in multiline scenarios, but  all of these searches are per-line.

Empact committed:
lint: Bump flake8 to 3.7.8

practicalswift commented:
utACK cf494074d84e5c31f8f614a91ab946f661b5a2d1

Thanks for doing this!

promag commented:
Concept ACK.

laanwj commented on contrib/devtools/copyright_header.py:
```diff
@@ -81,32 +81,32 @@ def compile_copyright_regex(copyright_style, year_style, name):
-    "Intel Corporation",
-    "The Zcash developers",
-    "Jeremy Rubin",
+    r"Satoshi Nakamoto\n",
```
As I commented on your previous request, turning these into raw strings is not correct. This is **not a pure refactor** and changes the meaning of the code.
```python
>>> r"Satoshi Nakamoto\n" == "Satoshi Nakamoto\n"
False
```

Raw strings are unnecessary here. If you want to avoid the warning stop escaping `.` as `\,`, e.g.
```patch
- "The LevelDB Authors\. All rights reserved\.\n",
+ "The LevelDB Authors\. All rights reserved.\n",
```

Empact replied:
If you run those strings against `re` instead you'll see something different. E.g. via `re.match`:
```
>>> import re
>>> re.match(".", "K") # standard string, no escape, matches any character
<re.Match object; span=(0, 1), match='K'>
>>> re.match("\.", "K") # standard string, with escape, does not match any character (hence no result)
>>> re.match("\.", ".") # standard string, with escape, does match "."
<re.Match object; span=(0, 1), match='.'>
>>> re.match(r"\.", "K") # raw string, with escape, does not match any character
>>> re.match(r"\.", ".") # raw string, with escape, does match "."
<re.Match object; span=(0, 1), match='.'>
>>> re.match("\\.", "K") # standard string, double escape, does not match any character
>>> re.match("\\.", ".") # standard string, double escape, does match "."
<re.Match object; span=(0, 1), match='.'>
>>> re.match(r"\\.", "K") # raw string, double escape, does not match any character
>>> re.match(r"\\.", ".") # raw string, double escape, does not match "."
```
You can see above that absent the `\`, "." is interpreted as any character, while r"\." and "\\." are both interpreted as an escaped period.

Empact replied:
Another way to look at this is comparing the script output with master. The output of `contrib/devtools/copyright_header.py report .` from project root currently matches between current master (https://github.com/bitcoin/bitcoin/commit/ab46fe6ec1b37e88c5a06ee7a06ab31cbd18304f) and this branch (https://github.com/bitcoin/bitcoin/pull/15257/commits/8f0b710f4f0e5087eba73cd047883353a25188ac) - but for one character, escaping the `.` in `'Wladimir J\. van der Laan\n'`, in keeping with the above, the matches are the same:
```
-------------------------------------------------------------------------------
667 files examined according to INCLUDE and EXCLUDE fnmatch rules
-------------------------------------------------------------------------------

  18 with zero copyrights
 504 with one copyright
 141 with two copyrights
   4 with three copyrights
   0 with four or more copyrights

-------------------------------------------------------------------------------
Copyrights with dominant style:
e.g. "Copyright (c)" and "<year>" or "<startYear>-<endYear>":

 135 with 'Satoshi Nakamoto\n'
 635 with 'The Bitcoin Core developers\n'
   1 with 'Bitcoin Core Developers\n'
   1 with 'University of Illinois at Urbana-Champaign\.\n'
   6 with 'Pieter Wuille\n'
   3 with 'Wladimir J\. van der Laan\n'
   3 with 'Jeff Garzik\n'
   1 with 'Jan-Klaas Kollhof\n'
   1 with 'Sam Rushing\n'
   2 with 'ArtForz -- public domain half-a-node\n'
   1 with 'Intel Corporation'
   6 with 'The Zcash developers'
   1 with 'Jeremy Rubin'

-------------------------------------------------------------------------------
Copyrights with year list style:
e.g. "Copyright (c)" and "<year1>, <year2>, ...":


-------------------------------------------------------------------------------
Copyrights with no "(c)" style:
e.g. "Copyright" and "<year>" or "<startYear>-<endYear>":

   1 with 'The Bitcoin Core developers\n'
   1 with 'BitPay Inc\.\n'

-------------------------------------------------------------------------------
0 with unexpected copyright holder names
-------------------------------------------------------------------------------
```

laanwj replied:
Okay then I was apparently misunderstanding what this is trying to do in the first place, sorry.

Empact replied:
All good, thanks!

Empact commented:
Rebased for #15258

practicalswift commented:
utACK 7a328c08485256bcb453fecf2cac02b0d53e8b57

Empact commented:
@MarcoFalke I pulled in your commit and it covered these additional cases: 59ffecf66cf4d08c4b431e457b083878d66a3fd6. The trouble is, it's not comprehensive and it's not easy to create a comprehensive scripted-diff given regex strings are not always directly connected with `re.` calls. I think limiting to addressing the flake warnings is reasonable given that.

Empact commented:
Rebased

MarcoFalke commented on contrib/devtools/copyright_header.py:
```diff
@@ -91,10 +91,10 @@ def compile_copyright_regex(copyright_style, year_style, name):
     "Pieter Wuille\n",
-    "Wladimir J. van der Laan\n",
+    r"Wladimir J\. van der Laan\n",
     "Jeff Garzik\n",
```
It seems inconsistent to have some lines without a newline, some with a '\n' and others with a '\\n'. I'd prefer to use the same suffix for all (or leave it and append the suffix at runtime)

Empact committed:
test/contrib: Fix invalid escapes in regex strings

Flagged by flake8 v3.6.0, as W605, plus a few others identified
incidentally, e.g. 59ffecf66cf4d08c4b431e457b083878d66a3fd6.

Note that r"\n" matches to "\n" under re.match/search.

JustinTArthur commented on contrib/devtools/copyright_header.py:
```diff
@@ -85,23 +85,23 @@ def get_filenames_to_examine(base_directory):
 
 def compile_copyright_regex(copyright_style, year_style, name):
-    return re.compile('%s %s,? %s' % (copyright_style, year_style, name))
+    return re.compile('%s %s,? %s\n' % (copyright_style, year_style, name))
```
Should this be a raw string literal now as well?

Empact replied:
It's not strictly required since the deprecation is about invalid escape sequences, and \n is valid. But the change in https://github.com/bitcoin/bitcoin/pull/15257/commits/45760242428fa53ba2719bbc35a68608cc4fc972 makes it necessary (because of `\*`).

laanwj commented:
Added 0.19 milestone, I have the feeling this going to need some iterations and thus too risky for 0.18.

Empact committed:
devtools: Accomodate block-style copyright blocks

Without this, `copyright_header.py report . verbose` reports:
-------------------------------------------------------------------------------
1 with unexpected copyright holder names
	./build_msvc/libsecp256k1_config.h
-------------------------------------------------------------------------------

Empact commented:
Rebased, bumped to 3.7.7, fixed additional lints.
Added another commit to fix a copyright detection in libsecp256k1_config.h.

Empact commented:
Rebased

practicalswift commented:
utACK efc39ea93990d1477988b0a9079afc788f64e3dc

Empact commented:
Rebased, bumped to 3.7.8.

practicalswift commented:
utACK 58054fb1c703f7602ef24470e37e44b4a9db4dbf

MarcoFalke commented on contrib/devtools/copyright_header.py:
```diff
@@ -41,8 +41,8 @@ def applies_to_file(filename):
-    return ((EXCLUDE_COMPILED.match(filename) is None) and
-            (INCLUDE_COMPILED.match(filename) is not None))
+    return ((EXCLUDE_COMPILED.match(filename) is None)
+            and (INCLUDE_COMPILED.match(filename) is not None))
```
`W504` should be disabled. This is not a critical error that should be blocking a merge

Empact committed:
lint: Disable flake8 W504 warning

In the words of MarcoFalke:
"W504 should be disabled. This is not a critical error that should be blocking a merge"
https://github.com/bitcoin/bitcoin/pull/15257#discussion_r302017280

Empact commented:
Rebased and disabled `W504`

practicalswift commented:
ACK 3d0a82cff8cbb809876e82dbe62d14d2adc07d94 -- diff looks correct
